File: cloud-functions/usertrades/exchanges/testnet/session.js
const ExchangeSession = require("../exchangeSession");
const CustomError = require("../../utils/error");
const { getTestnetPositions } = require("./positions");
const { getTestnetOrders, getTestnetOrderStatuses } = require("./orders");
const { getTestnetBalance } = require("./balance");

/**
 * Represents an exchange session for the testnet.
 * @extends ExchangeSession
 */
class TestnetSession extends ExchangeSession {
  /**
   * Creates a Testnet Session instance.
   * @param {string} apiKey - API key for the Testnet session.
   * @param {string} apiSecret - API secret for the Testnet session.
   */
  constructor(apiKey, apiSecret) {
    super(apiKey, apiSecret);
  }

  /**
   * Fetch orders for a given user ID.
   * @async
   * @param {string} userID - The unique ID of the user.
   * @returns {Promise<Array<Object>>} An array of order data objects.
   * @throws {CustomError} Throws a custom error if operation fails.
   */
  async getOrders(userID) {
    try {
      return await getTestnetOrders(this.apiKey, this.apiSecret, userID);
    } catch (e) {
      if (e instanceof CustomError) {
        throw e;
      }
      throw new CustomError({
        message: `Failed to fetch testnet orders: ${e.message}`,
        status: 500,
        source: "getOrders",
      });
    }
  }

  /**
   * Fetch positions for a given user ID.
   * @async
   * @param {string} userId - The unique ID of the user.
   * @returns {Promise<Array<Object>>} An array of position data objects.
   * @throws {CustomError} Throws a custom error if operation fails.
   */
  async getPositions(userId) {
    try {
      return await getTestnetPositions(this.apiKey, this.apiSecret, userId);
    } catch (e) {
      if (e instanceof CustomError) {
        throw e;
      }
      throw new CustomError({
        message: `Failed to fetch testnet positions: ${e.message}`,
        status: 500,
        source: "getPositions",
      });
    }
  }

  /**
   * Fetch the balance for a given trader ID.
   * @async
   * @returns {Promise<number>} The trader's balance.
   * @throws {CustomError} Throws a custom error if operation fails.
   */
  async getBalance() {
    try {
      return await getTestnetBalance(this.apiKey, this.apiSecret);
    } catch (e) {
      if (e instanceof CustomError) {
        throw e;
      }
      throw new CustomError({
        message: `Failed to fetch testnet balance: ${e.message}`,
        status: 500,
        source: "getBalance",
      });
    }
  }

  /**
   * Fetch order statuses for a given trader ID.
   * @async
   * @returns {Promise<Array<Object>>} An array of order status objects.
   * @throws {CustomError} Throws a custom error if operation fails.
   */
  async getOrderStatuses() {
    try {
      return await getTestnetOrderStatuses(this.apiKey, this.apiSecret);
    } catch (e) {
      if (e instanceof CustomError) {
        throw e;
      }
      throw new CustomError({
        message: `Failed to fetch testnet order statuses: ${e.message}`,
        status: 500,
        source: "getOrderStatuses",
      });
    }
  }
}

module.exports = TestnetSession;

File: cloud-run/solana-executor/exchanges/binance/api/session.py
from trade.exchange_session import ExchangeSession
from ..execution import bulk_order, cancel_all_orders, replace_sl, cancel_all_tps, bulk_tp, partial_close, send_sl, \
    cancel_order

class BinanceSession(ExchangeSession):
    """
    Represents a Binance exchange session.
    Extends ExchangeSession.
    """

    def __init__(self, api_key, api_secret):
        """
        Creates a BinanceSession instance.
        :param apiKey: API key for the Binance session.
        :param apiSecret: API secret for the Binance session.
        """
        super().__init__(api_key, api_secret)

    async def bulk_order(self, data):
        return await bulk_order(self.api_key, self.api_secret, data)

    async def cancel_all_orders(self, data):
        return await cancel_all_orders(self.api_key, self.api_secret, data)

    async def replace_sl(self, data):
        return await replace_sl(self.api_key, self.api_secret, data)

    async def cancel_all_tps(self, data):
        return await cancel_all_tps(self.api_key, self.api_secret, data)

    async def bulk_tp(self, data):
        return await bulk_tp(self.api_key, self.api_secret, data)

    async def partial_close(self, data):
        return await partial_close(self.api_key, self.api_secret, data)

    async def send_sl(self, data):
        return await send_sl(self.api_key, self.api_secret, data)

    async def cancel_order(self, data):
        return await cancel_order(self.api_key, self.api_secret, data)